File: RuckTracker/models/route.py
"""
Route model for AllTrails integration.
Represents shareable route data from various sources (AllTrails, custom, community).
"""

from dataclasses import dataclass
from typing import Optional, List, Dict, Any
from datetime import datetime
from decimal import Decimal

@dataclass
class RouteElevationPoint:
    """Individual elevation point along a route."""
    id: Optional[str]
    route_id: str
    distance_km: Decimal
    elevation_m: Decimal
    latitude: Optional[Decimal] = None
    longitude: Optional[Decimal] = None
    terrain_type: Optional[str] = None
    grade_percent: Optional[Decimal] = None
    created_at: Optional[datetime] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            'id': self.id,
            'route_id': self.route_id,
            'distance_km': float(self.distance_km),
            'elevation_m': float(self.elevation_m),
            'latitude': float(self.latitude) if self.latitude else None,
            'longitude': float(self.longitude) if self.longitude else None,
            'terrain_type': self.terrain_type,
            'grade_percent': float(self.grade_percent) if self.grade_percent else None,
            'created_at': self.created_at.isoformat() if self.created_at else None
        }

@dataclass
class RoutePointOfInterest:
    """Point of interest along a route (water, rest, viewpoint, etc.)."""
    id: Optional[str]
    route_id: str
    name: str
    description: Optional[str]
    poi_type: str  # water, rest, viewpoint, hazard, parking, landmark, shelter
    latitude: Decimal
    longitude: Decimal
    distance_from_start_km: Optional[Decimal] = None
    created_at: Optional[datetime] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        return {
            'id': self.id,
            'route_id': self.route_id,
            'name': self.name,
            'description': self.description,
            'poi_type': self.poi_type,
            'latitude': float(self.latitude),
            'longitude': float(self.longitude),
            'distance_from_start_km': float(self.distance_from_start_km) if self.distance_from_start_km else None,
            'created_at': self.created_at.isoformat() if self.created_at else None
        }

@dataclass
class Route:
    """Core route model for AllTrails integration."""
    
    # Core identification
    id: Optional[str]
    name: str
    description: Optional[str]
    source: str  # alltrails, custom, community
    external_id: Optional[str] = None  # AllTrails trail ID, etc.
    external_url: Optional[str] = None  # Link back to original source
    
    # Geographic data
    start_latitude: Decimal = None
    start_longitude: Decimal = None
    end_latitude: Optional[Decimal] = None
    end_longitude: Optional[Decimal] = None
    route_polyline: str = None  # Encoded polyline or GeoJSON
    
    # Route metrics
    distance_km: Decimal = None
    elevation_gain_m: Decimal = Decimal('0')
    elevation_loss_m: Optional[Decimal] = None
    min_elevation_m: Optional[Decimal] = None
    max_elevation_m: Optional[Decimal] = None
    
    # Difficulty and characteristics
    trail_difficulty: Optional[str] = None  # easy, moderate, hard, extreme
    trail_type: Optional[str] = None  # loop, out_and_back, point_to_point
    surface_type: Optional[str] = None  # trail, paved, gravel, mixed, rocky, technical
    
    # Popularity metrics
    total_planned_count: int = 0
    total_completed_count: int = 0
    average_rating: Optional[Decimal] = None
    
    # Metadata
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    created_by_user_id: Optional[str] = None
    is_verified: bool = False
    is_public: bool = True
    
    # Related data (loaded separately)
    elevation_points: Optional[List[RouteElevationPoint]] = None
    points_of_interest: Optional[List[RoutePointOfInterest]] = None

    def __post_init__(self):
        """Validate route data after initialization."""
        if self.source not in ['alltrails', 'custom', 'community']:
            raise ValueError(f"Invalid source: {self.source}")
        
        if self.trail_difficulty and self.trail_difficulty not in ['easy', 'moderate', 'hard', 'extreme']:
            raise ValueError(f"Invalid trail_difficulty: {self.trail_difficulty}")
        
        if self.trail_type and self.trail_type not in ['loop', 'out_and_back', 'point_to_point']:
            raise ValueError(f"Invalid trail_type: {self.trail_type}")
        
        if self.surface_type and self.surface_type not in ['trail', 'paved', 'gravel', 'mixed', 'rocky', 'technical']:
            raise ValueError(f"Invalid surface_type: {self.surface_type}")

    def to_dict(self, include_elevation_points: bool = False, include_pois: bool = False) -> Dict[str, Any]:
        """Convert to dictionary for JSON serialization."""
        data = {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'source': self.source,
            'external_id': self.external_id,
            'external_url': self.external_url,
            'start_latitude': float(self.start_latitude) if self.start_latitude else None,
            'start_longitude': float(self.start_longitude) if self.start_longitude else None,
            'end_latitude': float(self.end_latitude) if self.end_latitude else None,
            'end_longitude': float(self.end_longitude) if self.end_longitude else None,
            'route_polyline': self.route_polyline,
            'distance_km': float(self.distance_km) if self.distance_km else None,
            'elevation_gain_m': float(self.elevation_gain_m) if self.elevation_gain_m else None,
            'elevation_loss_m': float(self.elevation_loss_m) if self.elevation_loss_m else None,
            'min_elevation_m': float(self.min_elevation_m) if self.min_elevation_m else None,
            'max_elevation_m': float(self.max_elevation_m) if self.max_elevation_m else None,
            'trail_difficulty': self.trail_difficulty,
            'trail_type': self.trail_type,
            'surface_type': self.surface_type,
            'total_planned_count': self.total_planned_count,
            'total_completed_count': self.total_completed_count,
            'average_rating': float(self.average_rating) if self.average_rating else None,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            'created_by_user_id': self.created_by_user_id,
            'is_verified': self.is_verified,
            'is_public': self.is_public
        }
        
        # Optionally include related data
        if include_elevation_points and self.elevation_points:
            data['elevation_points'] = [point.to_dict() for point in self.elevation_points]
        
        if include_pois and self.points_of_interest:
            data['points_of_interest'] = [poi.to_dict() for poi in self.points_of_interest]
        
        return data

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'Route':
        """Create Route from dictionary (e.g., from database or API)."""
        # Convert string decimals back to Decimal objects for precision
        decimal_fields = [
            'start_latitude', 'start_longitude', 'end_latitude', 'end_longitude',
            'distance_km', 'elevation_gain_m', 'elevation_loss_m', 
            'min_elevation_m', 'max_elevation_m', 'average_rating'
        ]
        
        for field in decimal_fields:
            if data.get(field) is not None:
                data[field] = Decimal(str(data[field]))
        
        # Convert ISO datetime strings back to datetime objects
        if data.get('created_at'):
            data['created_at'] = datetime.fromisoformat(data['created_at'])
        if data.get('updated_at'):
            data['updated_at'] = datetime.fromisoformat(data['updated_at'])
        
        # Handle elevation points
        elevation_points = None
        if data.get('elevation_points'):
            elevation_points = [
                RouteElevationPoint(
                    id=point.get('id'),
                    route_id=point['route_id'],
                    distance_km=Decimal(str(point['distance_km'])),
                    elevation_m=Decimal(str(point['elevation_m'])),
                    latitude=Decimal(str(point['latitude'])) if point.get('latitude') else None,
                    longitude=Decimal(str(point['longitude'])) if point.get('longitude') else None,
                    terrain_type=point.get('terrain_type'),
                    grade_percent=Decimal(str(point['grade_percent'])) if point.get('grade_percent') else None,
                    created_at=datetime.fromisoformat(point['created_at']) if point.get('created_at') else None
                )
                for point in data['elevation_points']
            ]
        
        # Handle POIs
        points_of_interest = None
        if data.get('points_of_interest'):
            points_of_interest = [
                RoutePointOfInterest(
                    id=poi.get('id'),
                    route_id=poi['route_id'],
                    name=poi['name'],
                    description=poi.get('description'),
                    poi_type=poi['poi_type'],
                    latitude=Decimal(str(poi['latitude'])),
                    longitude=Decimal(str(poi['longitude'])),
                    distance_from_start_km=Decimal(str(poi['distance_from_start_km'])) if poi.get('distance_from_start_km') else None,
                    created_at=datetime.fromisoformat(poi['created_at']) if poi.get('created_at') else None
                )
                for poi in data['points_of_interest']
            ]
        
        # Remove nested data for main Route construction
        route_data = {k: v for k, v in data.items() if k not in ['elevation_points', 'points_of_interest']}
        
        route = cls(**route_data)
        route.elevation_points = elevation_points
        route.points_of_interest = points_of_interest
        
        return route

    def get_bounds(self) -> Optional[Dict[str, float]]:
        """Calculate geographic bounds for the route."""
        if not self.start_latitude or not self.start_longitude:
            return None
            
        # If we have elevation points with coordinates, use those for more accurate bounds
        if self.elevation_points:
            lats = [float(p.latitude) for p in self.elevation_points if p.latitude]
            lngs = [float(p.longitude) for p in self.elevation_points if p.longitude]
            
            if lats and lngs:
                return {
                    'north': max(lats),
                    'south': min(lats),
                    'east': max(lngs),
                    'west': min(lngs)
                }
        
        # Fallback to start/end points
        lats = [float(self.start_latitude)]
        lngs = [float(self.start_longitude)]
        
        if self.end_latitude and self.end_longitude:
            lats.append(float(self.end_latitude))
            lngs.append(float(self.end_longitude))
        
        return {
            'north': max(lats),
            'south': min(lats),
            'east': max(lngs),
            'west': min(lngs)
        }

    def is_loop_route(self) -> bool:
        """Check if this is a loop route based on start/end proximity."""
        if not all([self.start_latitude, self.start_longitude, self.end_latitude, self.end_longitude]):
            return self.trail_type == 'loop'
        
        # Calculate distance between start and end points (rough approximation)
        lat_diff = abs(float(self.start_latitude) - float(self.end_latitude))
        lng_diff = abs(float(self.start_longitude) - float(self.end_longitude))
        
        # If start and end are within ~100 meters, consider it a loop
        return lat_diff < 0.001 and lng_diff < 0.001

    def get_difficulty_score(self) -> int:
        """Get numeric difficulty score for calculations (1-4)."""
        difficulty_scores = {
            'easy': 1,
            'moderate': 2,
            'hard': 3,
            'extreme': 4
        }
        return difficulty_scores.get(self.trail_difficulty, 2)  # Default to moderate

File: rucking_app/lib/core/models/route.dart
import 'package:equatable/equatable.dart';
import 'route_elevation_point.dart';
import 'route_point_of_interest.dart';

/// Core route model for AllTrails integration
/// Represents a shareable route with geographic data, difficulty, and metadata
class Route extends Equatable {
  const Route({
    this.id,
    required this.name,
    this.description,
    required this.source,
    this.externalId,
    this.externalUrl,
    required this.routePolyline,
    required this.startLatitude,
    required this.startLongitude,
    this.endLatitude,
    this.endLongitude,
    required this.distanceKm,
    this.elevationGainM,
    this.elevationLossM,
    this.trailDifficulty,
    this.trailType,
    this.surfaceType,
    this.totalPlannedCount = 0,
    this.totalCompletedCount = 0,
    this.averageRating,
    this.isPublic = false,
    this.isVerified = false,
    this.createdByUserId,
    this.createdAt,
    this.updatedAt,
    this.elevationPoints = const [],
    this.pointsOfInterest = const [],
  });

  // Core identification
  final String? id;
  final String name;
  final String? description;
  final String source; // 'alltrails', 'custom', 'gpx_import', 'community'
  final String? externalId; // External route ID (e.g., AllTrails route ID)
  final String? externalUrl; // Link to original route source

  // Geographic data
  final String routePolyline; // Encoded polyline or coordinate string
  final double startLatitude;
  final double startLongitude;
  final double? endLatitude;
  final double? endLongitude;

  // Route metrics
  final double distanceKm;
  final double? elevationGainM;
  final double? elevationLossM;

  // Route characteristics
  final String? trailDifficulty; // 'easy', 'moderate', 'hard', 'extreme'
  final String? trailType; // 'loop', 'out_and_back', 'point_to_point'
  final String? surfaceType; // 'paved', 'dirt', 'gravel', 'mixed', 'varied'

  // Community data
  final int totalPlannedCount;
  final int totalCompletedCount;
  final double? averageRating; // 1-5 stars from user ratings

  // Privacy and verification
  final bool isPublic;
  final bool isVerified; // Verified routes from trusted sources

  // Metadata
  final String? createdByUserId;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  // Related data (loaded separately)
  final List<RouteElevationPoint> elevationPoints;
  final List<RoutePointOfInterest> pointsOfInterest;

  @override
  List<Object?> get props => [
        id,
        name,
        description,
        source,
        externalId,
        externalUrl,
        routePolyline,
        startLatitude,
        startLongitude,
        endLatitude,
        endLongitude,
        distanceKm,
        elevationGainM,
        elevationLossM,
        trailDifficulty,
        trailType,
        surfaceType,
        totalPlannedCount,
        totalCompletedCount,
        averageRating,
        isPublic,
        isVerified,
        createdByUserId,
        createdAt,
        updatedAt,
        elevationPoints,
        pointsOfInterest,
      ];

  /// Create Route from API JSON response
  factory Route.fromJson(Map<String, dynamic> json) {
    return Route(
      id: json['id'] as String?,
      name: json['name']?.toString() ?? 'Unknown Route',
      description: json['description'] as String?,
      source: json['source']?.toString() ?? 'unknown',
      externalId: json['external_id'] as String?,
      externalUrl: json['external_url'] as String?,
      routePolyline: json['route_polyline']?.toString() ?? '',
      startLatitude: json['start_latitude'] != null ? (json['start_latitude'] as num).toDouble() : 0.0,
      startLongitude: json['start_longitude'] != null ? (json['start_longitude'] as num).toDouble() : 0.0,
      endLatitude: json['end_latitude'] != null 
          ? (json['end_latitude'] as num).toDouble() 
          : null,
      endLongitude: json['end_longitude'] != null 
          ? (json['end_longitude'] as num).toDouble() 
          : null,
      distanceKm: (json['distance_km'] as num).toDouble(),
      elevationGainM: json['elevation_gain_m'] != null 
          ? (json['elevation_gain_m'] as num).toDouble() 
          : null,
      elevationLossM: json['elevation_loss_m'] != null 
          ? (json['elevation_loss_m'] as num).toDouble() 
          : null,
      trailDifficulty: json['trail_difficulty'] as String?,
      trailType: json['trail_type'] as String?,
      surfaceType: json['surface_type'] as String?,
      totalPlannedCount: json['total_planned_count'] as int? ?? 0,
      totalCompletedCount: json['total_completed_count'] as int? ?? 0,
      averageRating: json['average_rating'] != null 
          ? (json['average_rating'] as num).toDouble() 
          : null,
      isPublic: json['is_public'] as bool? ?? false,
      isVerified: json['is_verified'] as bool? ?? false,
      createdByUserId: json['created_by_user_id'] as String?,
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at'] as String) 
          : null,
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at'] as String) 
          : null,
      elevationPoints: json['elevation_points'] != null
          ? (json['elevation_points'] as List)
              .map((e) => RouteElevationPoint.fromJson(e as Map<String, dynamic>))
              .toList()
          : [],
      pointsOfInterest: json['points_of_interest'] != null
          ? (json['points_of_interest'] as List)
              .map((e) => RoutePointOfInterest.fromJson(e as Map<String, dynamic>))
              .toList()
          : [],
    );
  }

  /// Convert Route to JSON for API requests
  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'name': name,
      if (description != null) 'description': description,
      'source': source,
      if (externalId != null) 'external_id': externalId,
      if (externalUrl != null) 'external_url': externalUrl,
      'route_polyline': routePolyline,
      'start_latitude': startLatitude,
      'start_longitude': startLongitude,
      if (endLatitude != null) 'end_latitude': endLatitude,
      if (endLongitude != null) 'end_longitude': endLongitude,
      'distance_km': distanceKm,
      if (elevationGainM != null) 'elevation_gain_m': elevationGainM,
      if (elevationLossM != null) 'elevation_loss_m': elevationLossM,
      if (trailDifficulty != null) 'trail_difficulty': trailDifficulty,
      if (trailType != null) 'trail_type': trailType,
      if (surfaceType != null) 'surface_type': surfaceType,
      'total_planned_count': totalPlannedCount,
      'total_completed_count': totalCompletedCount,
      if (averageRating != null) 'average_rating': averageRating,
      'is_public': isPublic,
      'is_verified': isVerified,
      if (createdByUserId != null) 'created_by_user_id': createdByUserId,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
      if (elevationPoints.isNotEmpty) 
        'elevation_points': elevationPoints.map((e) => e.toJson()).toList(),
      if (pointsOfInterest.isNotEmpty) 
        'points_of_interest': pointsOfInterest.map((e) => e.toJson()).toList(),
    };
  }

  /// Create a copy of this Route with updated fields
  Route copyWith({
    String? id,
    String? name,
    String? description,
    String? source,
    String? externalId,
    String? externalUrl,
    String? routePolyline,
    double? startLatitude,
    double? startLongitude,
    double? endLatitude,
    double? endLongitude,
    double? distanceKm,
    double? elevationGainM,
    double? elevationLossM,
    String? trailDifficulty,
    String? trailType,
    String? surfaceType,
    int? totalPlannedCount,
    int? totalCompletedCount,
    double? averageRating,
    bool? isPublic,
    bool? isVerified,
    String? createdByUserId,
    DateTime? createdAt,
    DateTime? updatedAt,
    List<RouteElevationPoint>? elevationPoints,
    List<RoutePointOfInterest>? pointsOfInterest,
  }) {
    return Route(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      source: source ?? this.source,
      externalId: externalId ?? this.externalId,
      externalUrl: externalUrl ?? this.externalUrl,
      routePolyline: routePolyline ?? this.routePolyline,
      startLatitude: startLatitude ?? this.startLatitude,
      startLongitude: startLongitude ?? this.startLongitude,
      endLatitude: endLatitude ?? this.endLatitude,
      endLongitude: endLongitude ?? this.endLongitude,
      distanceKm: distanceKm ?? this.distanceKm,
      elevationGainM: elevationGainM ?? this.elevationGainM,
      elevationLossM: elevationLossM ?? this.elevationLossM,
      trailDifficulty: trailDifficulty ?? this.trailDifficulty,
      trailType: trailType ?? this.trailType,
      surfaceType: surfaceType ?? this.surfaceType,
      totalPlannedCount: totalPlannedCount ?? this.totalPlannedCount,
      totalCompletedCount: totalCompletedCount ?? this.totalCompletedCount,
      averageRating: averageRating ?? this.averageRating,
      isPublic: isPublic ?? this.isPublic,
      isVerified: isVerified ?? this.isVerified,
      createdByUserId: createdByUserId ?? this.createdByUserId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      elevationPoints: elevationPoints ?? this.elevationPoints,
      pointsOfInterest: pointsOfInterest ?? this.pointsOfInterest,
    );
  }

  // Computed properties and helper methods

  /// Get difficulty level as enum for easier handling
  RouteDifficulty get difficultyLevel {
    switch (trailDifficulty?.toLowerCase()) {
      case 'easy':
        return RouteDifficulty.easy;
      case 'moderate':
        return RouteDifficulty.moderate;
      case 'hard':
        return RouteDifficulty.hard;
      case 'extreme':
        return RouteDifficulty.extreme;
      default:
        return RouteDifficulty.moderate;
    }
  }

  /// Get trail type as enum for easier handling
  RouteType get routeType {
    switch (trailType?.toLowerCase()) {
      case 'loop':
        return RouteType.loop;
      case 'out_and_back':
        return RouteType.outAndBack;
      case 'point_to_point':
        return RouteType.pointToPoint;
      default:
        return RouteType.outAndBack;
    }
  }

  /// Get formatted distance string
  String get formattedDistance {
    if (distanceKm < 1) {
      return '${(distanceKm * 1000).toStringAsFixed(0)}m';
    } else {
      return '${distanceKm.toStringAsFixed(1)}km';
    }
  }

  /// Get formatted elevation gain string
  String get formattedElevationGain {
    if (elevationGainM == null) return 'Unknown elevation';
    return '${elevationGainM!.toStringAsFixed(0)}m elevation';
  }

  /// Get popularity score based on usage
  double get popularityScore {
    if (totalCompletedCount == 0) return 0.0;
    
    double score = totalCompletedCount * 10.0;
    
    // Bonus for high rating
    if (averageRating != null && averageRating! >= 4.0) {
      score += averageRating! * 5;
    }
    
    // Bonus for verified routes
    if (isVerified) {
      score += 10;
    }
    
    return score.clamp(0.0, 100.0);
  }

  /// Check if route has elevation profile data
  bool get hasElevationProfile => elevationPoints.isNotEmpty;

  /// Check if route has points of interest
  bool get hasPointsOfInterest => pointsOfInterest.isNotEmpty;

  /// Get estimated duration in minutes based on difficulty and distance
  int get estimatedDurationMinutes {
    // Base pace in minutes per km
    double pacePerKm = switch (difficultyLevel) {
      RouteDifficulty.easy => 12.0,
      RouteDifficulty.moderate => 15.0,
      RouteDifficulty.hard => 18.0,
      RouteDifficulty.extreme => 22.0,
    };
    
    // Adjust for elevation gain
    if (elevationGainM != null && elevationGainM! > 0) {
      double elevationAdjustment = (elevationGainM! / 100) * 2; // 2 minutes per 100m elevation
      pacePerKm += elevationAdjustment;
    }
    
    return (distanceKm * pacePerKm).round();
  }

  /// Get formatted estimated duration
  String get formattedEstimatedDuration {
    final minutes = estimatedDurationMinutes;
    if (minutes < 60) {
      return '${minutes}min';
    } else {
      final hours = minutes ~/ 60;
      final remainingMinutes = minutes % 60;
      if (remainingMinutes == 0) {
        return '${hours}h';
      } else {
        return '${hours}h ${remainingMinutes}min';
      }
    }
  }

  /// Check if user can access this route
  bool canUserAccess(String? userId) {
    return isPublic || createdByUserId == userId;
  }

  /// Get source display name
  String get sourceDisplayName {
    switch (source.toLowerCase()) {
      case 'alltrails':
        return 'AllTrails';
      case 'custom':
        return 'Custom Route';
      case 'gpx_import':
        return 'GPX Import';
      case 'community':
        return 'Community';
      default:
        return source;
    }
  }
}

/// Route difficulty levels
enum RouteDifficulty {
  easy,
  moderate,
  hard,
  extreme;

  String get displayName {
    switch (this) {
      case RouteDifficulty.easy:
        return 'Easy';
      case RouteDifficulty.moderate:
        return 'Moderate';
      case RouteDifficulty.hard:
        return 'Hard';
      case RouteDifficulty.extreme:
        return 'Extreme';
    }
  }

  String get description {
    switch (this) {
      case RouteDifficulty.easy:
        return 'Suitable for beginners';
      case RouteDifficulty.moderate:
        return 'Some fitness required';
      case RouteDifficulty.hard:
        return 'Challenging route';
      case RouteDifficulty.extreme:
        return 'Very challenging';
    }
  }
}

/// Route types
enum RouteType {
  loop,
  outAndBack,
  pointToPoint;

  String get displayName {
    switch (this) {
      case RouteType.loop:
        return 'Loop';
      case RouteType.outAndBack:
        return 'Out & Back';
      case RouteType.pointToPoint:
        return 'Point to Point';
    }
  }
}
